From 1f0d79abc4b82dbed010e83cd4094d66b942061c Mon Sep 17 00:00:00 2001
From: Surya Muttamsetty <muttamsetty.surya@tcs.com>
Date: Tue, 9 May 2017 12:37:42 +0530
Subject: [PATCH 1/2] EXT-334 OXS Aggregate Flow Stats Message Handling

---
 lib/ofp-util.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/lib/ofp-util.c b/lib/ofp-util.c
index a261d98..681b648 100644
--- a/lib/ofp-util.c
+++ b/lib/ofp-util.c
@@ -3282,6 +3282,12 @@ ofputil_encode_aggregate_stats_reply(
     enum ofpraw raw;
 
     ofpraw_decode(&raw, request);
+    if(raw==OFPRAW_OFPST15_OXS_AGGREGATE_REQUEST) {
+    msg = ofpraw_alloc_stats_reply(request, 0);
+    enum ofp_version version = request->version;
+    oxs_put_agg_stat(msg,stats,version);
+    }
+    else {
     if (raw == OFPRAW_OFPST10_AGGREGATE_REQUEST) {
         packet_count = unknown_to_zero(stats->packet_count);
         byte_count = unknown_to_zero(stats->byte_count);
@@ -3295,6 +3301,7 @@ ofputil_encode_aggregate_stats_reply(
     put_32aligned_be64(&asr->packet_count, htonll(packet_count));
     put_32aligned_be64(&asr->byte_count, htonll(byte_count));
     asr->flow_count = htonl(stats->flow_count);
+    }
 
     return msg;
 }
@@ -3304,13 +3311,23 @@ ofputil_decode_aggregate_stats_reply(struct ofputil_aggregate_stats *stats,
                                      const struct ofp_header *reply)
 {
     struct ofpbuf msg = ofpbuf_const_initializer(reply, ntohs(reply->length));
-    ofpraw_pull_assert(&msg);
+    enum ofperr error;
+    enum ofpraw raw;
 
+    raw=ofpraw_pull_assert(&msg);
+    if (raw == OFPRAW_OFPST15_OXS_AGGREGATE_REPLY){
+      memset(stats, 0, sizeof(*stats));
+      error =  oxs_pull_agg_stat(msg, stats);
+      if(error)
+        return error;
+    }
+    else{
     struct ofp_aggregate_stats_reply *asr = msg.msg;
     stats->packet_count = ntohll(get_32aligned_be64(&asr->packet_count));
     stats->byte_count = ntohll(get_32aligned_be64(&asr->byte_count));
     stats->flow_count = ntohl(asr->flow_count);
 
+    }
     return 0;
 }
 
-- 
2.1.0

